include/ua_securitypolicy.h

```diff
@@ -0,0 +1,241 @@
+     * refer to the OPC UA specification 1.03 Part6 Table 33
+     *
+     * \param secret
+     * \param seed
+     * \param out an output to write the data to. The length defines the maximum
+     *            number of output bytes that are produced.
+     */
+    UA_StatusCode (*const generateKey)(const UA_SecurityPolicy *const securityPolicy,
+                                       const UA_ByteString *const secret,
+                                       const UA_ByteString *const seed,
+                                       UA_ByteString *const out);
+    /**
+     * \brief Random generator for generating nonces.
+     * 
+     * Generates a nonce.
+     * 
+     * \param securityPolicy the securityPolicy this function is invoked on. Example: myPolicy->generateNonce(myPolicy, &outBuff);
+     * \param out pointer to a buffer to store the nonce in. Needs to be allocated by the caller.
+     *            The buffer is filled with random data.
+     */
+    UA_StatusCode (*const generateNonce)(const UA_SecurityPolicy *const securityPolicy,
+                                         const void *const endpointContext,
+                                         UA_ByteString *const out);
+
+    /**
+     * \brief Calculates the padding size for a message with the specified amount of bytes.
+     *
+     * \param securityPolicy the securityPolicy the function is invoked on.
+     * \param bytesToWrite the size of the payload plus the sequence header, since both need to be encoded
+     * \param paddingSize out parameter. Will contain the paddingSize byte.
+     * \param extraPaddingSize out parameter. Will contain the extraPaddingSize. If no extra padding is needed, this is 0.
+     * \return the total padding size consisting of high and low bytes.
+     */
+    UA_UInt16 (*const calculatePadding)(const UA_SecurityPolicy *const securityPolicy,
+                                        const size_t bytesToWrite,
+                                        UA_Byte *const paddingSize,
+                                        UA_Byte *const extraPaddingSize);
+
+    const UA_SecurityPolicySigningModule signingModule;
+
+    const size_t signingKeyLength;
+    const size_t encryptingKeyLength;
+    const size_t encryptingBlockSize;
+} UA_SecurityPolicySymmetricModule;
+
+/**
+ * \brief This struct describes a security policy. All functions and modules need to be implemented.
+ */
+struct UA_SecurityPolicy {
+    /* The policy uri that identifies the implemented algorithms */
+    const UA_ByteString policyUri;
+
+    /**
+     * \brief Verifies the certificate using the trust list and revocation list in the policy context
+     *
+     * \param policyContext the policy context that contains the revocation and trust lists.
+     * \param channelContext the channel context that contains the already parsed certificate.
+     */
+    UA_StatusCode (*const verifyCertificate)(const UA_SecurityPolicy *const securityPolicy,
```

Comment:
It could be done in one function, so that we don't allocate memory for the context before verifying the certificate, but that would mean having three different functionalities (or two depending on how you look at it) in one function: init, then certificate parsing, then verification (or any other order??)
An alternative would be to have parseRemoteCertificate return a void pointer to some context that holds the certificate. This gets passed to verify and then we can create the channelContext. After that the certificate would be set to that context with some function like
`channelContext.setRemoteCertificate(......)`

I find the latter approach somewhat better.
What's your take on this?